sdk/go/dagger/client/connhelper/engineconn/engineconn.go

```diff
@@ -0,0 +1,95 @@
+package engineconn
+
+import (
+	"context"
+	"io"
+	"net"
+	"net/http"
+	"sync"
+	"time"
+
+	"go.dagger.io/dagger/engine"
+)
+
+type EngineConn struct {
+	stopCh chan struct{}
+	doneCh chan error
+}
+
+func New() *EngineConn {
+	return &EngineConn{
+		stopCh: make(chan struct{}),
+		doneCh: make(chan error),
+	}
+}
+
+func (c *EngineConn) Connect(ctx context.Context) (*http.Client, error) {
+	started := make(chan struct{})
+	var client *http.Client
+	go func() {
+		defer close(c.doneCh)
+		err := engine.Start(ctx, nil, func(ctx engine.Context) error {
+			// TODO: not efficient, but whatever
+			serverConn, clientConn := net.Pipe()
+
+			l := &singleConnListener{
+				conn: serverConn,
+			}
+			s := http.Server{
+				Handler:           ctx.Handler,
+				ReadHeaderTimeout: 30 * time.Second,
+			}
+			go s.Serve(l)
+
+			client = &http.Client{
+				Transport: &http.Transport{
+					DialContext: func(_ context.Context, _, _ string) (net.Conn, error) {
+						return clientConn, nil
+					},
+				},
+			}
+			close(started)
+			<-c.stopCh
+			return nil
+		})
+		c.doneCh <- err
+	}()
+
+	select {
+	case <-started:
+		return client, nil
+	case err := <-c.doneCh:
+		return client, err
+	}
+}
+
+func (c *EngineConn) Close() error {
+	close(c.stopCh)
```

Comment:
nit: if a user accidentally called `Close()` multiple times, I think they'd get a panic due to `close(c.stopCh)` being called multiple times, should probably just be no-op after the first Close instead.